### ISSUE TITLE
sqlfluff fix with dbt templater replaces aliases with sources

### ISSUE THREAD

**Issue by tekumara** [sqlfluff/sqlfluff#2051]
## Expected Behaviour

No replacement of aliases with the source, because the source database and schema changes between environments.

## Observed Behaviour
```
SELECT t.dv_RepositoryID
FROM {{ source('mydb', 'repo') }} t
```
is replaced with
```
SELECT mydb_dev.repo.dv_repositoryid
FROM {{ source('mydb', 'repo') }} t
```

However `mydb_dev` is the development database.

## Dialect
snowflake

## Version
sqlfluff, version 0.8.2
python 3.9

## Configuration

```
[tool.sqlfluff.core]
templater = "dbt"
sql_file_exts = ".sql,.sql.j2,.dml,.ddl"
dialect = "snowflake"
ignore_templated_areas = true

[tool.sqlfluff.indentation]
indented_joins = false
indented_using_on = true
template_blocks_indent = false

[tool.sqlfluff.templater.dbt]
project_dir = "."
profile = "snowflake"
```

**Comment by jpy-git** [comment]
@tekumara what lint was being raised by the original query? Helps narrow things down 😄

**Comment by tekumara** [comment]
```
L:   2 | P:  60 | L031 | Avoid aliases in from clauses and join conditions.
```

**Comment by jpy-git** [comment]
This makes sense thanks.

So L031 forbids the use of aliases in FROM/JOIN clauses
i.e. instead of
```sql
SELECT t.foo
FROM bar AS t
```
it suggests you should use
```sql
SELECT bar.foo
FROM bar
```
and provides an autofix to correct this.

What's happening in this case is the same but the table name is a templated variable. So in sqlfluff we initially substitute in the templated variables, evaluate the rules, and then substitute back the templates and we have some logic to not apply rules if they impact a templated area as that can corrupt the SQL.
What's different in this case is the fix is using the value from the templated area and applying it elsewhere in the query, hence not triggering our logic to ignore the fix.

I think that L031 needs some extra logic to abandon the fix if the table name it uses is from a template.

**Comment by jpy-git** [comment]
I've tested this with Jinja and can confirm I see the same thing so this is a general templating issue.

MRE:
*test.sql*
```sql
SELECT t.foo
FROM {{ tablename }} AS t
```
*.sqlfluff*
```
[sqlfluff:templater:jinja:context]
tablename=bar
```

Run `sqlfluff fix test.sql`

![Screen-Recording-2021-12-07-at-10 09 53](https://user-images.githubusercontent.com/80432516/145012009-0586f91b-2bcc-4569-a61d-7210d1f46b6e.gif)

**Comment by barrywhart** [comment]
Interesting issue. I agree that L031 will need changes, but let's try and think about a more general, long-term fix.

My thoughts: when a rule is copying content from one place to another, this should be captured as part of the LintFix object somehow. With this information, we can enhance the logic added in #2117  to prevent copying content out of a templated area.

This could be adding position info to the new/replacement segment(s) or a new field on LintFix. I think a new field will be better, because using position on the segments is potentially ambiguous (sometimes rules move/reuse existing segments, making it hard to tell if the position info is a leftover or a deliberately set value).
